File: 2022/day03-2.hs
import Data.Char (ord)

solution :: String -> Int
solution = sum . map (priority . badge) . groupBy . lines

badge :: (String, String, String) -> Char
badge (x : xs, b, c)
  | x `elem` b && x `elem` c = x
  | otherwise = badge (xs, b, c)

groupBy :: [String] -> [(String, String, String)]
groupBy (a : b : c : xs) = (a, b, c) : groupBy xs
groupBy [] = []

priority :: Char -> Int
priority c
  | o >= ord 'a' && o <= ord 'z' = o - ord 'a' + 1
  | o >= ord 'A' && o <= ord 'Z' = o - ord 'A' + 27
  | otherwise = 0
  where
    o = ord c

main :: IO ()
main = do readFile "day03.dat" >>= print . solution

File: 2022/day03_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return sum(priority(badge(item)) for item in group_by(content))


def badge(group: tuple[str, str, str]):
    fst, snd, trd = group
    return next(ch for ch in fst if ch in snd and ch in trd)


def group_by(lines: list[str]):
    rest = lines
    output = []
    while len(rest) > 0:
        fst, snd, trd, *rest = rest
        output.append((fst, snd, trd))
    return output


def priority(ch: str):
    if "a" <= ch[0] <= "z":
        return ord(ch) - ord("a") + 1
    if "A" <= ch[0] <= "Z":
        return ord(ch) - ord("A") + 27
    return 0


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
